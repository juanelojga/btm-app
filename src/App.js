import React, {Component} from 'react'
import PropTypes from 'prop-types'
import find from 'lodash/find'
import debounce from 'lodash/debounce'

import Suggestions from './components/Suggestions'
import ShowResult from './components/ShowResult'

import './App.scss'

const DEBOUNCE_DELAY = 600

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {name: '', users: [], selectedName: '', showSuggestions: false, isLoading: false, isFocussed: false}
    this.searchDebounced = debounce(this.search, DEBOUNCE_DELAY)
  }

  handleChange = e => {
    const {minNumberOfChars} = this.props
    const value = e.target.value

    this.setState({name: value}, () => {
      if (value.length >= minNumberOfChars) {
        this.searchDebounced(value)
      }
    })
  }

  handleSubmit = e => {
    e.preventDefault()
    const {name} = this.state
    this.setState({selectedName: name})
  }

  handleClick = value => {
    const {users} = this.state
    const user = find(users, ['id', value])
    this.setState({name: user.name, showSuggestions: false})
  }

  handleClickOutside = () => {
    this.setState({showSuggestions: false})
  }

  handleOnFocus = () => {
    this.setState({isFocussed: true})
  }

  handleOnBlur = () => {
    this.setState({isFocussed: false})
  }

  search = q => {
    this.setState({isLoading: true, showSuggestions: true}, () => {
      this.props.api.users.get({q}).then(users => {
        const {name} = this.state
        const showSuggestions = name.length >= this.props.minNumberOfChars
        this.setState({users, showSuggestions, isLoading: false})
      })
    })
  }

  render() {
    const inputClassName = ['form-control']
    if (this.state.isFocussed) inputClassName.push('show-focus')

    return (
      <>
        <form onSubmit={this.handleSubmit} className="form-wrapper">
          <div className={inputClassName.join(' ')}>
            <label>
              Name
              <input
                type="text"
                value={this.state.name}
                onChange={this.handleChange}
                onFocus={this.handleOnFocus}
                onBlur={this.handleOnBlur}
              />
            </label>
          </div>
          {this.state.showSuggestions ? (
            <Suggestions
              items={this.state.users}
              onClick={this.handleClick}
              onClickOutside={this.handleClickOutside}
              isLoading={this.state.isLoading}
            />
          ) : (
            ''
          )}

          <input type="submit" value="Submit" className="button" disabled={!this.state.name} />
        </form>
        <ShowResult value={this.state.selectedName} />
      </>
    )
  }
}

App.defaultProps = {
  minNumberOfChars: 3,
}

App.propTypes = {
  minNumberOfChars: PropTypes.number,
  api: PropTypes.object,
}

export default App
