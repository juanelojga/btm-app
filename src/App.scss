*,
*::after,
*::before {
  margin: 0;
  padding: 0;
  box-sizing: inherit;
}

html {
  font-size: 62.5%; //1 rem = 10px; 10px/16px = 62.5%
}

body {
  box-sizing: border-box;
  padding: 5rem;
  background-color: #f1f4f5;
  font-size: 1.4rem;
}

.form-wrapper {
  position: relative;
  .form-control {
    display: block;
    padding: 0.5rem;
    border: 0.1rem solid #a0a0a1;
    margin: 2rem 0 0;
    width: 30rem;
    background-color: #fff;
    &.show-focus {
      border-color: #0098d0;
      label {
        color: #0098d0;
      }
    }
    input {
      outline: none;
      border: none;
      display: block;
      line-height: 1.2%;
      width: 100%;
      padding: 0.5rem;
    }
    label {
      display: block;
      font-size: 0.8rem;
      color: #a0a0a1;
    }
  }
  .button {
    background-color: #0098d0;
    border: 0.1rem solid #0098d0;
    border-radius: 0.3rem;
    color: #fff;
    padding: 1rem 2rem;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    margin: 20px 0;
    &:disabled {
      background-color: #fff;
      border-color: #a0a0a1;
      color: #a0a0a1;
    }
  }
}
