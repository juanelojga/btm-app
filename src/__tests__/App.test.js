import React from 'react'
import {render, fireEvent, wait, cleanup} from 'react-testing-library'
import faker from 'faker'

import App from '../App'

afterEach(cleanup)

test('calls users get when input has 3 or more characters', async () => {
  const value = 'example'
  const fakeApi = {
    users: {
      get: jest.fn(() => Promise.resolve()),
    },
  }
  const {getByLabelText} = render(<App api={fakeApi} />)

  const name = getByLabelText('Name')
  fireEvent.change(name, {target: {value}})

  await wait(() => {
    expect(fakeApi.users.get).toHaveBeenCalledTimes(1)
    expect(fakeApi.users.get).toHaveBeenCalledWith({q: value})
  })
})

test('not calls users get when input less than 3 characters', async () => {
  const value = 'ex'
  const fakeApi = {
    users: {
      get: jest.fn(() => Promise.resolve()),
    },
  }
  const {getByLabelText} = render(<App api={fakeApi} />)

  const name = getByLabelText('Name')
  fireEvent.change(name, {target: {value}})

  await wait(() => {
    expect(fakeApi.users.get).not.toHaveBeenCalled()
  })
})

test('submit button disabled if input is empty', () => {
  const fakeApi = {
    users: {
      get: jest.fn(() => Promise.resolve()),
    },
  }
  const {getByText} = render(<App api={fakeApi} />)

  const submit = getByText('Submit')
  expect(submit.disabled).toBeTruthy()
})

test('submit button enabled if input is not empty', () => {
  const value = faker.name.findName()
  const fakeApi = {
    users: {
      get: jest.fn(() => Promise.resolve()),
    },
  }

  const {getByText, getByLabelText} = render(<App api={fakeApi} />)

  const name = getByLabelText('Name')
  fireEvent.change(name, {target: {value}})

  const submit = getByText('Submit')
  expect(submit.disabled).not.toBeTruthy()
})

test('snapshot', () => {
  const {container} = render(<App />)
  expect(container).toMatchSnapshot()
})
