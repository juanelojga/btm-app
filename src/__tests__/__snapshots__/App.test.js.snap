// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <form
    class="form-wrapper"
  >
    <div
      class="full-input"
    >
      <label>
        Name
        <input
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      class="loading"
    />
    <div
      class="list"
    />
    <input
      class="button"
      disabled=""
      type="submit"
      value="Submit"
    />
  </form>
  <div>
    Submitted Value: 
    
  </div>
</div>
`;
